lib/block/blockdownload.js

```diff
@@ -0,0 +1,223 @@
+            console.log("Block download: block download controller starting file download");
+            this.emit("filestart", transferEvent.transferAsset.eventData);
+        });
+
+        controller.on(TransferEvents.JOIN_TRANSFER_PARTS, transferEvent => {
+            this.emit("fileprogress", {
+                ...transferEvent.transferAsset.eventData,
+                transferred: transferEvent.props.transferBytes
+            });
+        });
+
+        controller.on(TransferEvents.AFTER_JOIN_TRANSFER_PARTS, transferEvent => {
+            console.log("Block download: block download controller finishing file download");
+            this.emit("fileend", transferEvent.transferAsset.eventData);
+        });
+
+        controller.on(TransferEvents.ERROR, transferEvent => {
+            console.log(`Error during block download: ${transferEvent.error}`);
+            this.errorEvents.push(transferEvent);
+
+            if (transferEvent.props.firstError) {
+                this.emit("fileerror", {
+                    ...transferEvent.transferAsset.eventData,
+                    errors: [DownloadError.fromError(transferEvent.error)]
+                });
+            }
+        });
+
+        console.log("Block download: block download controller created");
+        return controller;
+    }
+
+    /**
+     * Removes all listeners for events 
+     * `CREATE_TRANSFER_PARTS`, `JOIN_TRANSFER_PARTS`,
+     * `AFTER_JOIN_TRANSFER_PARTS` and `ERROR`
+     * from a block download controller
+     * @param {TransferController} controller a block download controller
+     */
+    finalizeController(controller){
+        if(!controller){
+            return;
+        }
+
+        if (this.errorEvents) {
+            this.errorEvents = null;
+        }
+        
+        try {
+            controller.removeAllListeners(TransferEvents.CREATE_TRANSFER_PARTS);
+            controller.removeAllListeners(TransferEvents.JOIN_TRANSFER_PARTS);
+            controller.removeAllListeners(TransferEvents.AFTER_JOIN_TRANSFER_PARTS);
+            controller.removeAllListeners(TransferEvents.ERROR);
+        } catch (err) {
+            console.log(`Failed to remove event listeners from block download controller: ${err}`);
+        }
+    }
+
+    /**
+     * Download files to local disk
+     * Throws the first unrecoverable error if unsuccessful, all others are logged
+     * 
+     * @param {BlockDownloadOptions} options Block download options
+     */
+    async downloadFiles(options = {}) {
+        const preferredPartSize = options.preferredPartSize || DEFAULT_PART_SIZE;
+        const maxConcurrent = options.maxConcurrent || DEFAULT_MAX_CONCURRENCY;
+
+        // Build and execute pipeline
+        const controller = this.createBlockDownloadController();
+        const randomFileAccess = new RandomFileAccess();
+        try {
+            const pipeline = new Pipeline(
+                new GetAssetMetadata(options),
```

Comment:
we should avoid the GetAssetMetadata call if we already have the needed metadata. Those can then be set [while creating the TransferAsset](https://github.com/adobe/node-httptransfer/blob/feature/nui-1644/lib/block/blockdownload.js#L56)

Reply:
+1 and we will always have this data when nui is the client since we do a head request in core

Reply:
Changed. Now it requests metadata only if any of the needed ones are missing.